### PR TITLE
[15.0][FIX] purchase_discount_*: use hook method for supplierinfo

### PR DESCRIPTION
**Pull request by chienandalu**
From version 15 we've got this handy _prepare_supplier_info method so we don't have to do context injection juggling to add the discount info.

please review @victoralmau @sergio-teruel

### CODE REVIEW

**Review by sergio-teruel — APPROVED** [review]

**Review by victoralmau — APPROVED** [review]

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
It's better to not put 2 return in the method.

If there's no `vals["product_tmpl_id"]`, then you should check `vals["product_tmpl_id"]` and extract the template from it.

**Comment by chienandalu** [comment]
> It's better to not put 2 return in the method.
> 
> If there's no `vals["product_tmpl_id"]`, then you should check `vals["product_tmpl_id"]` and extract the template from it.

The thing is supplierinfo records can be created either without any product info or without this context or both.

**Comment by pedrobaeza** [comment]
Correct, but you still need to know the product template through the mechanism I'm talking about.

**Comment by pedrobaeza** [comment]
@chienandalu I have rebased resolving conflicts (due to a new `.get("product_tmpl_id")` instead of `["product_tmpl_id"]`). Can you please check my comment to finish this?

**Comment by chienandalu** [comment]
> If there's no vals["product_tmpl_id"], then you should check vals["product_tmpl_id"] and extract the template from it.

Do you mean from vals["product_id"]?

**Comment by pedrobaeza** [comment]
Yes, sorry.

**Comment by pedrobaeza** [comment]
@chienandalu can you please finish this?

**Comment by chienandalu** [comment]
The main issue was already fixed here: https://github.com/OCA/purchase-workflow/pull/2079

Anyway, I implemented your request, although I'm not completly sure it's needed for the case of this context (the creation of the supplierinfo from the method `_add_supplier_to_product`, which should be quite predictable

File: purchase_discount/models/product_supplierinfo.py [hunk 1]
```diff
@@ -27,9 +27,19 @@ def _get_po_to_supplierinfo_synced_fields(self):
     def create(self, vals_list):
         """Insert discount (or others) from context from purchase.order's
         _add_supplier_to_product method"""
-        for vals in vals_list:
-            product_tmpl_id = vals.get("product_tmpl_id")
-            po_line_map = self.env.context.get("po_line_map", {})
+        po_line_map = self.env.context.get("po_line_map")
+        if not po_line_map:
```
Comment (pedrobaeza) [inline purchase_discount/models/product_supplierinfo.py]: I don't understand what this exit condition does.

File: purchase_discount/models/product_supplierinfo.py [hunk 2]
```diff
@@ -27,9 +27,19 @@ def _get_po_to_supplierinfo_synced_fields(self):
...
+            return super().create(vals_list)
+        for vals in filter(
+            lambda x: x.get("product_tmpl_id", x.get("product_id")), vals_list
```
Comment (pedrobaeza) [inline purchase_discount/models/product_supplierinfo.py]: This filter seems now incorrect if we fallback to the other value.

File: purchase_discount/models/product_supplierinfo.py (hunk 1 above)
Comment (chienandalu) [reply purchase_discount/models/product_supplierinfo.py]: Well, I was in my way to explain how this is a dict injected in context with the only purpose of passing the discount values on creation when I realized that in this version there's a hook method that makes this engineering unnecessary

**Review by pedrobaeza — APPROVED** [review]
This is way better!

/ocabot merge patch

File: purchase_discount/models/product_supplierinfo.py (hunk 2 above)
Comment (chienandalu) [reply purchase_discount/models/product_supplierinfo.py]: Not important now, but it just meant that we could go with that dictionary as we could get the template from the product_id

**Comment by rousseldenis** [comment]
@chienandalu Do you plan to forwardport this in 16 ?

**Comment by chienandalu** [comment]
The thing is that the methods changed completely and now there isn't an obvious hook, so maybe an extra PR to Odoo will be needed to try solve it...

**Comment by rousseldenis** [comment]
> The thing is that the methods changed completely and now there isn't an obvious hook, so maybe an extra PR to Odoo will be needed to try solve it...

Ok. I'll maybe look at it

**Comment by chienandalu** [comment]
Great! ping me if you want to when you're on it :smiley:

**Pull request merged**